### PR TITLE
Add support for creating volume and network with label definition

### LINKED ISSUE #3892: Proposal: Support for creating volume and network with label options
I have ran into a situation where I need to get networks and volumes created by a given compose project. I think it could be done by adding a label to the networks and volumes defined in the compose file and then I could query the networks and volumes by a label filter through docker api. 

But when  I check the schema definition at **compose/config/config_schema_v2.0.json**, I found it's not supported by compose.

Would it be better if creating volume and network with label options is supported? I could do this if this feature is reasonable.

### PR DESCRIPTION
**Pull request by dbdd4us**
Fixes #3892 

**- What I did**
Add volume and network label support for compose file v2

**- How I did it**
1. Add **labels** to volume and network section in **config_schema_v2.0.json** 
2. Add implementation
3. Add doc of volume and network label to **docs/compose-file.md**

**-How to verify it**
Add test in **tests/unit/config/config_test.py**, **tests/integration/project_test.py** and **tests/acceptance/cli_test.py**

@shin- PTAL

Signed-off-by: dbdd <wangtong2712@gmail.com>

### CODE REVIEW

File: tests/integration/project_test.py [hunk 1]
```diff
@@ -810,6 +810,46 @@ def test_project_up_volumes(self):
...
+        )
+
+        project.up()
+
+        volumes = [
+            v for v in self.client.volumes().get('Volumes', [])
+            if v['Name'].startswith('composetest_')
+        ]
+
+        assert sorted(v['Name'] for v in volumes) == ['composetest_{}'.format(volume_name)]
```
Comment (shin-) [inline tests/integration/project_test.py]: nit: we don't need `sorted()` here since there's a single element

File: tests/acceptance/cli_test.py [hunk 2]
```diff
@@ -755,6 +755,26 @@ def test_up_with_external_default_network(self):
...
+        self.dispatch(['-f', filename, 'up', '-d'], returncode=0)
+
+        volume_with_label = '{}_volume_with_label'.format(self.project.name)
+
+        volumes = [
+            v for v in self.client.volumes().get('Volumes', [])
+            if v['Name'].startswith('{}_'.format(self.project.name))
+        ]
+
+        assert sorted(v['Name'] for v in volumes) == [volume_with_label]
```
Comment (shin-) [inline tests/acceptance/cli_test.py]: nit: we don't need `sorted()` here since there's a single element

**Comment by shin-** [comment]
Thanks! This is quite thourough. There's a few things I see:

1. I have one minor nitpick on the test code, see above.
2. We need to update the docker-py dependency in `requirements.txt`. We can use `235964607d200b57ea41383e54473c2976f4d730`
3. Since this requires API 1.23, we may actually want to / have to make it part of a `2.1` compose file spec. It's actually something we've already done for another option in #3653 , but just hasn't been merged yet. cc @dnephin @aanand on this.

**Comment by dbdd4us** [comment]
PR updated. PTAL @shin- 

1. Fixes nits
2. Update docker-py dependency to `235964607d200b57ea41383e54473c2976f4d730`

Do I need to use `@v2_1_only` in the releated test code and add `[Version 2.1 file format]` to the releated doc?

**Comment by alexmavr** [comment]
@dbdd4us are you planning to address the network part of  https://github.com/docker/compose/issues/3892 as well? This PR only seems to address volume labels

**Comment by dbdd4us** [comment]
@alex

**Comment by dbdd4us** [comment]
PR updated. PTAL @alexmavr @shin-

1. Add network part
2. Update docker-py dependency to 1.10.2

I was wondering if I need to use `@v2_1_only` in the releated test code and add [Version 2.1 file format] to the releated doc? or do these things after when #3653 is merged?

File: compose/config/config_schema_v2.0.json [hunk 3]
```diff
@@ -248,7 +248,8 @@
           },
           "additionalProperties": false
         },
-        "internal": {"type": "boolean"}
+        "internal": {"type": "boolean"},
+        "labels": {"$ref": "#/definitions/list_or_dict"}
```
Comment (shin-) [inline compose/config/config_schema_v2.0.json]: Now that config_schema_v2.1.json has been merged, we can move this property there.

File: compose/config/config_schema_v2.0.json [hunk 4]
```diff
@@ -270,6 +271,7 @@
             "name": {"type": "string"}
           }
         },
+        "labels": {"$ref": "#/definitions/list_or_dict"},
```
Comment (shin-) [inline compose/config/config_schema_v2.0.json]: Now that config_schema_v2.1.json has been merged, we can move this property there.

File: requirements.txt [hunk 5]
```diff
@@ -1,7 +1,7 @@
 PyYAML==3.11
 backports.ssl-match-hostname==3.5.0.1; python_version < '3'
 cached-property==1.2.0
-docker-py==1.9.0
+docker-py==1.10.2
```
Comment (shin-) [inline requirements.txt]: You'll probably get a merge conflict here since we bumped to 1.10.3 in master.

File: tests/acceptance/cli_test.py [hunk 6]
```diff
@@ -755,6 +755,46 @@ def test_up_with_external_default_network(self):
         assert list(container.get('NetworkSettings.Networks')) == [network_name]
 
     @v2_only()
```
Comment (shin-) [inline tests/acceptance/cli_test.py]: Update this to `@v2_1_only()`

File: tests/acceptance/cli_test.py [hunk 7]
```diff
@@ -755,6 +755,46 @@ def test_up_with_external_default_network(self):
...
+        networks = [
+            n for n in self.client.networks()
+            if n['Name'].startswith('{}_'.format(self.project.name))
+        ]
+
+        assert [n['Name'] for n in networks] == [network_with_label]
+
+        assert networks[0]['Labels'] == {'label_key': 'label_val'}
+
+    @v2_only()
```
Comment (shin-) [inline tests/acceptance/cli_test.py]: Update this to `@v2_1_only()`

File: tests/integration/project_test.py [hunk 8]
```diff
@@ -784,6 +784,42 @@ def test_project_up_with_network_internal(self):
         assert network['Internal'] is True
 
     @v2_only()
```
Comment (shin-) [inline tests/integration/project_test.py]: Update this to `@v2_1_only()`

File: tests/integration/project_test.py [hunk 9]
```diff
@@ -810,6 +846,46 @@ def test_project_up_volumes(self):
         self.assertEqual(volume_data['Driver'], 'local')
 
     @v2_only()
```
Comment (shin-) [inline tests/integration/project_test.py]: Update this to `@v2_1_only()`

File: docs/compose-file.md [hunk 10]
```diff
@@ -833,6 +833,22 @@ refer to it within the Compose file:
         external:
           name: actual-name-of-volume
 
+### labels
+
```
Comment (shin-) [inline docs/compose-file.md]: Needs the `> Added in version 2.1 file format` line, similar to [this](https://github.com/docker/compose/blob/master/docs/compose-file.md#link_local_ips)

File: docs/compose-file.md [hunk 11]
```diff
@@ -932,6 +948,22 @@ refer to it within the Compose file:
         external:
           name: actual-name-of-network
 
+### labels
+
```
Comment (shin-) [inline docs/compose-file.md]: Needs the `> Added in version 2.1 file format` line, similar to [this](https://github.com/docker/compose/blob/master/docs/compose-file.md#link_local_ips)

**Comment by dbdd4us** [comment]
@shin- PR Updated, PTAL 

I'm confused about the `Reorder python imports Failed` in Jenkins build.

**Comment by dnephin** [comment]
"Reorder python imports" is a lint/style check performed by pre-commit. It means the imports are not ordered correctly. You can either install http://pre-commit.com/ and run it to fix the issue, or you can re-order them manually.

Imports should be sorted and grouped according to "built-in", "third-party", "local"

**Comment by dbdd4us** [comment]
@dnephin 
Thanks, it has been fixed. PR Updated, PTAL : )

**Comment by shin-** [comment]
retest this please

**Comment by shin-** [comment]
WIth the docs moving to a separate repository, this needs a rebase that removes the `docs/compose-file.md`. @dbdd4us , do you mind taking care of it? Thanks, and sorry about the delay on merging this.

**Comment by dbdd4us** [comment]
@shin- 
Rebased, but it seems that jenkins ci is broken, I'm not sure if it's accidental.

**Comment by allencloud** [comment]
LGTM (IANAM)
Any Update?
Thanks

**Comment by bwowk** [comment]
Closes #3107

**Pull request merged**